software/base/src/main/java/brooklyn/entity/service/SystemServiceEnricher.java

```diff
@@ -0,0 +1,152 @@
+    static final Set<String> LAUNCH_EFFECTOR_NAMES = ImmutableSet.of("start", "restart");
+    public static final ConfigKey<String> LAUNCH_SCRIPT_NAME = ConfigKeys.newStringConfigKey(
+            "service.script_name", "The name of the launch script to be created in the runtime directory of the entity.", "service-launch.sh");
+    public static final ConfigKey<String> SERVICE_NAME = ConfigKeys.newStringConfigKey(
+            "service.name", "The name of the system service. Can use ${entity_name} and ${id} variables to template the value.", "${entity_name}-${id}");
+
+    @Override
+    public void setEntity(EntityLocal entity) {
+        super.setEntity(entity);
+        subscribeLaunch();
+    }
+
+    private void subscribeLaunch() {
+        if(checkCanSubscribe()) {
+            getSubscriptionTracker().subscribe(entity, Attributes.SERVICE_STATE_ACTUAL, new EntityLaunchListener(this));
+        }
+    }
+
+    public void onLaunched(Task<?> task) {
+        WrappedStream streamStdin = BrooklynTaskTags.stream(task, BrooklynTaskTags.STREAM_STDIN);
+        if (streamStdin == null) return;
+
+        WrappedStream streamEnv = BrooklynTaskTags.stream(task, BrooklynTaskTags.STREAM_ENV);
+        String stdin = streamStdin.streamContents.get();
+        String env = streamEnv.streamContents.get();
+
+        final SshMachineLocation sshMachine = EffectorTasks.getSshMachine(entity);
+        final String launchScriptPath = Urls.mergePaths(getRunDir(), getStartScriptName());
+
+        Task<Void> installerTask = TaskBuilder.<Void>builder()
+                .name("install (service)")
+                .description("Install as a system service")
+                .body(new Runnable() {
+                    @Override
+                    public void run() {
+                        ProcessTaskFactory<Integer> taskFactory = SshTasks.newSshExecTaskFactory(sshMachine, "[ -e '" + launchScriptPath + "' ]")
+                                .summary("check installed")
+                                .allowingNonZeroExitCode();
+                        boolean isInstalled = DynamicTasks.queue(taskFactory).get() == 0;
+                        if (!isInstalled) {
+                            Task<?> serviceInstallTask = SystemServiceInstallerFactory.forMachine(entity, SystemServiceEnricher.this).getServiceInstallTask();
+                            DynamicTasks.queue(serviceInstallTask);
+                        }
+                    }
+                })
+                .build();
+
+        SshPutTaskWrapper updateLaunchScriptTask = SshTasks.newSshPutTaskFactory(sshMachine, launchScriptPath).contents(getLaunchScript(stdin, env)).newTask();
+        ProcessTaskWrapper<Integer> makeExecutableTask = SshTasks.newSshExecTaskFactory(sshMachine, "chmod +x " + launchScriptPath)
+            .requiringExitCodeZero()
+            .newTask();
+        Task<Void> udpateTask = TaskBuilder.<Void>builder()
+            .name("update-launch")
```

Comment:
Personal preference for 8 char indent (to follow oracle/google coding conventions), as is done on lines 109-115 below. I particularly like consistency within a class.